### ISSUE TITLE
UUID converstion?

### ISSUE THREAD

**Issue by b-jsshapiro** [AnatolyUss/nmig#78]
In MySQL, UUID values are often stored as char(36) or binary(16). In postgres there is a native UUID type. Any way to implement this conversion?

In our case, we already have a database function UuidFromBin() that would convert the binary(16) values back to UUID strings, so if there were some way in data_types_map.json to say "wrap this column value with the named conversion function during the SELECT", that would be very helpful and perhaps not very hard to implement.

I haven't looked at the code, so it might perhaps be necessary to also say the output type returned by the conversion function.

So the model would be: given an input column `C` of type `MySQL:T1`, fetch the column value using SELECT expression `F(C) as C` to get type `MySQL:T2`, and then convert type `MySQL:T2` to target type `Posgres:OutputType` using the current rules.

I am sure that other places would need to be touched, but UUID conversion seems important, and different people use different rearrangements when creating the binary(16) values.

I suppose an alternative would be to create computed columns prior to migration, but since UUID is often the primary key this starts to get mixed up with foreign constraints, so it is not straightforward. The nice thing about saying "I'm just converting through an intermediate step" is that all of the constraints and indexes have a clean mapping.

Any chance something like this may be possible? Is there already a way to do it that I am failing to consider?

Thanks!

**Comment by AnatolyUss** [comment]
Hello Jonathan,
The <code>data_types_map.json</code> is used across all tables and all columns without exceptions. 
Once you modify one of the mappings, it will be applied to all appropriate columns in all tables.
How about to change the column's type to <code>UUID</code> on migration completion?
Would it solve the issue?
Please, have a look on the following snippet:
<code>create table test_1(id bigserial, id_char36 char(36), id_uuid uuid, primary key(id));</code>
<code>insert into test_1(id_char36, id_uuid) values('dba077d2-2c2a-11eb-9eec-a8e47dc668b4', 'dba077d2-2c2a-11eb-9eec-a8e47dc668b4');</code>
<code>select * from test_1;</code>
<code>alter table test_1 alter column id_char36 type uuid using id_char36::uuid;</code>
<code>select * from test_1;</code>

**Comment by b-jsshapiro** [comment]
Yes. Or at least all of the tables and columns for that configuration.

We will be doing this migration several times during testing. Every manual step is an opportunity for errors. I was able to edit the code locally to add conversions char(36)->uuid and binary(16)->uuid. The latter relies on invoking a stored procedure UuidFromBin() in the mysql database at the time of select, so the procedure deal with any permutations that were done for efficiency.

Having added those, I added the necessary entries to `data_types_map.json`.

UUID conversion is actually a very common case. If you are willing, I would suggest adopting the edit (it's simple), adding the entries to `data_types_map.json`, and then _commenting them out_ by default. That way they won't end up getting applied unless somebody has enabled them on purpose, but they are present for users who need to make this very common migration change. Let me know if you would like that as a pull request.

In the course of doing that change, I looked at the way the mapping is being done, and I actually think it would benefit from a modest rearrangement. I've been working on a GraphQL to SQL query planner lately, which needed a similar kind of conversion model, so the issues are fresh in my mind. I spent some time looking at the code to see what it would take. The pieces are all there, but way the configurations are currently handled makes the change harder than it needs to be. If you have any interest in it as an idea for the future, I can write it up as an issue.

Shalom
